src/liballoc/allocator.rs

```diff
@@ -0,0 +1,786 @@
+    ///
+    /// On arithmetic overflow, returns `None`.
+    pub fn repeat_packed(&self, n: usize) -> Option<Self> {
+        let scaled = match self.size().checked_mul(n) {
+            None => return None,
+            Some(scaled) => scaled,
+        };
+        let size = { assert!(scaled > 0); scaled };
+        Some(Layout { size: size, align: self.align })
+    }
+
+    /// Creates a layout describing the record for `self` followed by
+    /// `next` with no additional padding between the two. Since no
+    /// padding is inserted, the alignment of `next` is irrelevant,
+    /// and is not incoporated *at all* into the resulting layout.
+    ///
+    /// Returns `(k, offset)`, where `k` is layout of the concatenated
+    /// record and `offset` is the relative location, in bytes, of the
+    /// start of the `next` embedded witnin the concatenated record
+    /// (assuming that the record itself starts at offset 0).
+    ///
+    /// (The `offset` is always the same as `self.size()`; we use this
+    ///  signature out of convenience in matching the signature of
+    ///  `fn extend`.)
+    ///
+    /// On arithmetic overflow, returns `None`.
+    pub fn extend_packed(&self, next: Self) -> Option<(Self, usize)> {
+        let new_size = match self.size().checked_add(next.size()) {
+            None => return None,
+            Some(new_size) => new_size,
+        };
+        Some((Layout { size: new_size, ..*self }, self.size()))
+    }
+
+    // Below family of methods *assume* inputs are pre- or
+    // post-validated in some manner. (The implementations here
+    ///do indirectly validate, but that is not part of their
+    /// specification.)
+    //
+    // Since invalid inputs could yield ill-formed layouts, these
+    // methods are `unsafe`.
+
+    /// Creates layout describing the record for a single instance of `T`.
+    pub unsafe fn new_unchecked<T>() -> Self {
+        let (size, align) = size_align::<T>();
+        Layout::from_size_align(size, align)
+    }
+
+
+    /// Creates a layout describing the record for `self` followed by
+    /// `next`, including any necessary padding to ensure that `next`
+    /// will be properly aligned. Note that the result layout will
+    /// satisfy the alignment properties of both `self` and `next`.
+    ///
+    /// Returns `(k, offset)`, where `k` is layout of the concatenated
+    /// record and `offset` is the relative location, in bytes, of the
+    /// start of the `next` embedded witnin the concatenated record
+    /// (assuming that the record itself starts at offset 0).
+    ///
+    /// Requires no arithmetic overflow from inputs.
+    pub unsafe fn extend_unchecked(&self, next: Self) -> (Self, usize) {
+        self.extend(next).unwrap()
+    }
+
+    /// Creates a layout describing the record for `n` instances of
+    /// `self`, with a suitable amount of padding between each.
+    ///
+    /// Requires non-zero `n` and no arithmetic overflow from inputs.
```

Comment:
We have to make a choice about where we are going to attempt to enforce that people do not make zero-sized memory requests to the allocator.

Given how many iterations the API for Layout and Allocator went through, I can believe that at different times I had different models in my head about whose responsibility it was to ensure that the underlying system allocator did not receive requests for zero-sized memory...

anyway, I'll try to resolve this. not 100% sure which way I'll go (I don't think I have a clear "right direction" from prior discussions on the RFC thread...)